### PR TITLE
Allow for case insensitive header matching

### PR DESCRIPTION
**Pull request by davinci26**
Similar to query params allow for case insensitive case matching.
    
Signed-off-by: Sotiris Nanopoulos <sotiris.nanopoulos@reddit.com>

### CODE REVIEW

**Comment by davinci26** [comment]
@clayton-gonsalves wanna do the first-round of review?

File: internal/envoy/v3/route.go
Comment (davinci26) [inline internal/envoy/v3/route.go:804]: this could implemented as `contains` api in Envoy as well, happy to make that change

Comment (clayton-gonsalves) [reply internal/envoy/v3/route.go:804]: We should create an issue to track this so that it doesn't get lost.

File: apis/projectcontour/v1/httpproxy.go
Comment (clayton-gonsalves) [inline apis/projectcontour/v1/httpproxy.go:125]: should also add this note at the top level `HeaderMatchCondition struct`

**Review by clayton-gonsalves — CHANGES_REQUESTED** [review]
lgtm modulo 2 minor comments

File: internal/envoy/v3/route.go
Comment (davinci26) [reply internal/envoy/v3/route.go:804]: done: https://github.com/projectcontour/contour/issues/5572

File: apis/projectcontour/v1/httpproxy.go
Comment (davinci26) [reply apis/projectcontour/v1/httpproxy.go:126]: fixed in 33758a5d52a01e106dbef383f0b49f11b235c0b9

**Comment by davinci26** [comment]
@clayton-gonsalves PTAL

**Review by clayton-gonsalves — APPROVED** [review]

**Comment by clayton-gonsalves** [comment]
hey @skriss, this LGTM would be nice to a get a one if the maintainers 👍 . Thanks

**Comment by davinci26** [comment]
Hey, @skriss can I have a review please :D I am porting a fairly large haproxy config to contour and there are 3 missing features from the routing I wanted to contribute to Contour and this PR adds a bunch of tests that make it much easier to test the rest, so i would like to get it a path to be merged.

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks @davinci26 (and @clayton-gonsalves for the review as well), this looks pretty good, just a couple small things. 

Also, please update `site/content/docs/main/config/request-routing.md` to include the new field (can more or less copy from the query param section)

File: apis/projectcontour/v1/httpproxy.go
Comment (davinci26) [reply apis/projectcontour/v1/httpproxy.go]: fixed in c0470d1d82c14549e3a2f40304709266055a082c

File: changelogs/unreleased/5567-davinci26-small.md
Comment (davinci26) [reply changelogs/unreleased/5567-davinci26-small.md]: fixed in c0470d1d82c14549e3a2f40304709266055a082c

**Comment by davinci26** [comment]
There is a test that is failing but it has passed in 9f48c616dc6307afd42f784ed6a3145be122dc00 and then I have only comment changes so I am speculating it is a flaky test rather than code issue. If it goes red again I will investigate as it will be blocking the PR

File: apis/projectcontour/v1/httpproxy.go
Comment (skriss) [inline apis/projectcontour/v1/httpproxy.go:92]:
```suggestion
// Contains, NotContains, Exact, NotExact and Regex can be set.
```

**Comment by davinci26** [comment]
`commit suggestion breaks DCO` will need to force push

**Comment by skriss** [comment]
> There is a test that is failing but it has passed in https://github.com/projectcontour/contour/commit/9f48c616dc6307afd42f784ed6a3145be122dc00 and then I have only comment changes so I am speculating it is a flaky test rather than code issue. If it goes red again I will investigate as it will be blocking the PR

Gateway-conformance is currently pretty flaky, we're eagerly awaiting the release of v0.8.0 which will have some fixes to help. In the mean time we'll re-run as needed, doesn't look like anything related to your changes.

**Comment by davinci26** [comment]
@skriss is there a command to retry a single test?

**Comment by davinci26** [comment]
ty for the reviews folks!

**Pull request merged**
